JsonFx/JsonFx.Json/JsonReader.cs

```diff
@@ -644,11 +644,46 @@ private IEnumerable ReadArray(Type arrayType)
 
 			// TODO: optimize to reduce number of conversions on lists
 
-			if (arrayItemType != null && arrayItemType != typeof(object))
+			if (arrayItemType != null)
 			{
-				// if all items are of same type then convert to array of that type
-				return jsArray.ToArray(arrayItemType);
-			}
+				if (arrayItemType == typeof(object[]))
+                {
+                    bool hasObjectArray = false;
+                    bool hasNonObjectArray = false;
+                    foreach (var obj in jsArray)
+                    {
+                        if (obj.GetType() == typeof(object[]))
+                        {
+							hasObjectArray = true;
+                        }
+						else if (obj.GetType().IsArray)
+                        {
+							hasNonObjectArray = true;
+                        }
+                    }
+
+					// Special handling for converting an object[] that contains both an object[] and a non-object[].
```

Comment:
"non-object[]" do you mean: an array of another element type?
this way it kind of looks like it could be a mixture of `object[]` and `float` or something. 
oh god. I just find myself thinking: what if we had `[[{},{}],null]` - it might be worth a test? 🤷

Reply:
Yep exactly that. we want to deal with arrays that contain both `object[]` and `float[]` (which should produce `object[][]`)
I will add some more tests, you've suggested some good cases